File: ios/objNative/JSBufferAnalystParameter.m
//
//  JSBufferAnalystParameter.m
//  Supermap
//
//  Created by 王子豪 on 2017/5/17.
//  Copyright © 2017年 Facebook. All rights reserved.
//

#import "JSBufferAnalystParameter.h"

#import "SuperMap/BufferAnalystParameter.h"
#import "JSObjManager.h"
@implementation JSBufferAnalystParameter
RCT_EXPORT_MODULE();
RCT_REMAP_METHOD(createObj,createObjWithResolver:(RCTPromiseResolveBlock)resolve rejecter:(RCTPromiseRejectBlock)reject){
    @try {
        BufferAnalystParameter* para = [[BufferAnalystParameter alloc]init];
        NSInteger key = (NSInteger)para;
        [JSObjManager addObj:para];
        resolve(@{@"bufferAnalystParameterId":@(key).stringValue});
    } @catch (NSException *exception) {
        reject(@"BufferAnalystParameter",@"create object exception",nil);
    }
}

RCT_REMAP_METHOD(setEndType,setEndTypeById:(NSString*)paraId endType:(int)endType resolver:(RCTPromiseResolveBlock)resolve rejecter:(RCTPromiseRejectBlock)reject){
    @try {
        BufferAnalystParameter* para = [JSObjManager getObjWithKey:paraId];
        para.bufferEndType = endType;
        resolve(@"endType setted");
    } @catch (NSException *exception) {
        reject(@"BufferAnalystParameter",@"set EndType exception",nil);
    }
}

RCT_REMAP_METHOD(getEndType,getEndTypeById:(NSString*)paraId resolver:(RCTPromiseResolveBlock)resolve rejecter:(RCTPromiseRejectBlock)reject){
    @try {
        BufferAnalystParameter* para = [JSObjManager getObjWithKey:paraId];
        int endType = para.bufferEndType;
        NSNumber* typeNum = [NSNumber numberWithInt:endType];
        resolve(@{@"EndType":typeNum});
    } @catch (NSException *exception) {
        reject(@"BufferAnalystParameter",@"get EndType exception",nil);
    }
}

RCT_REMAP_METHOD(setLeftDistance,setLeftDistanceById:(NSString*)paraId distance:(double)distance resolver:(RCTPromiseResolveBlock)resolve rejecter:(RCTPromiseRejectBlock)reject){
    @try {
        BufferAnalystParameter* para = [JSObjManager getObjWithKey:paraId];
        para.leftDistance = [NSString stringWithFormat:@"%f",distance];
        resolve(@"left Distance setted");
    } @catch (NSException *exception) {
        reject(@"BufferAnalystParameter",@"set left Distance exception",nil);
    }
}

RCT_REMAP_METHOD(setLeftDistanceByStr,setLeftDistanceByStringAndId:(NSString*)paraId distance:(NSString*)distance resolver:(RCTPromiseResolveBlock)resolve rejecter:(RCTPromiseRejectBlock)reject){
    @try {
        BufferAnalystParameter* para = [JSObjManager getObjWithKey:paraId];
        para.leftDistance = distance;
        resolve(@"left Distance setted");
    } @catch (NSException *exception) {
        reject(@"BufferAnalystParameter",@"set left Distance exception",nil);
    }
}

RCT_REMAP_METHOD(getLeftDistance,getLeftDistanceById:(NSString*)paraId resolver:(RCTPromiseResolveBlock)resolve rejecter:(RCTPromiseRejectBlock)reject){
    @try {
        BufferAnalystParameter* para = [JSObjManager getObjWithKey:paraId];
        NSString* distance = para.leftDistance;
        resolve(@{@"leftDistance":distance});
    } @catch (NSException *exception) {
        reject(@"BufferAnalystParameter",@"get left Distance exception",nil);
    }
}

RCT_REMAP_METHOD(setRightDistance,setRightDistanceById:(NSString*)paraId distance:(double)distance resolver:(RCTPromiseResolveBlock)resolve rejecter:(RCTPromiseRejectBlock)reject){
    @try {
        BufferAnalystParameter* para = [JSObjManager getObjWithKey:paraId];
        para.rightDistance = [NSString stringWithFormat:@"%f",distance];
        resolve(@"right Distance setted");
    } @catch (NSException *exception) {
        reject(@"BufferAnalystParameter",@"set right Distance exception",nil);
    }
}

RCT_REMAP_METHOD(setRightDistanceByStr,setRightDistanceByStrAndId:(NSString*)paraId distance:(NSString*)distance resolver:(RCTPromiseResolveBlock)resolve rejecter:(RCTPromiseRejectBlock)reject){
    @try {
        BufferAnalystParameter* para = [JSObjManager getObjWithKey:paraId];
        para.rightDistance = distance;
        resolve(@"left Distance setted");
    } @catch (NSException *exception) {
        reject(@"BufferAnalystParameter",@"set left Distance exception",nil);
    }
}

RCT_REMAP_METHOD(getRightDistance,getRightDistanceById:(NSString*)paraId resolver:(RCTPromiseResolveBlock)resolve rejecter:(RCTPromiseRejectBlock)reject){
    @try {
        BufferAnalystParameter* para = [JSObjManager getObjWithKey:paraId];
        NSString* distance = para.rightDistance;
        resolve(@{@"rightDistance":distance});
    } @catch (NSException *exception) {
        reject(@"BufferAnalystParameter",@"get right Distance exception",nil);
    }
}

RCT_REMAP_METHOD(setRadiusUnit,setRadiusUnitById:(NSString*)paraId radiusUnit:(int)radiusUnit resolver:(RCTPromiseResolveBlock)resolve rejecter:(RCTPromiseRejectBlock)reject){
    @try {
        BufferAnalystParameter* para = [JSObjManager getObjWithKey:paraId];
        para.bufferRadiusUnit = radiusUnit;
        resolve(@"RadiusUnit setted");
    } @catch (NSException *exception) {
        reject(@"BufferAnalystParameter",@"set Radius Unit exception",nil);
    }
}

RCT_REMAP_METHOD(getRadiusUnit,getRadiusUnitById:(NSString*)paraId resolver:(RCTPromiseResolveBlock)resolve rejecter:(RCTPromiseRejectBlock)reject){
    @try {
        BufferAnalystParameter* para = [JSObjManager getObjWithKey:paraId];
        int unit = para.bufferRadiusUnit;
        NSNumber* nsUnit = [NSNumber numberWithInt:unit];
        resolve(@{@"radiusUnit":nsUnit});
    } @catch (NSException *exception) {
        reject(@"BufferAnalystParameter",@"get Radius Unit exception",nil);
    }
}

RCT_REMAP_METHOD(setSemicircleLineSegment,setSemicircleLineSegmentById:(NSString*)paraId segment:(int)segment resolver:(RCTPromiseResolveBlock)resolve rejecter:(RCTPromiseRejectBlock)reject){
    @try {
        BufferAnalystParameter* para = [JSObjManager getObjWithKey:paraId];
        para.semicircleLineSegment = segment;
        resolve(@"SemicircleLineSegment setted");
    } @catch (NSException *exception) {
        reject(@"BufferAnalystParameter",@"set Semicircle Line Segment exception",nil);
    }
}

RCT_REMAP_METHOD(getSemicircleLineSegment,getSemicircleLineSegmentById:(NSString*)paraId resolver:(RCTPromiseResolveBlock)resolve rejecter:(RCTPromiseRejectBlock)reject){
    @try {
        BufferAnalystParameter* para = [JSObjManager getObjWithKey:paraId];
        int segment = para.semicircleLineSegment;
        NSNumber* nsSegment = [NSNumber numberWithInt:segment];
        resolve(@{@"segment":nsSegment});
    } @catch (NSException *exception) {
        reject(@"BufferAnalystParameter",@"get Semicircle Line Segment exception",nil);
    }
}
@end

File: android/src/main/java/com/supermap/rnsupermap/JSBufferAnalystParameter.java
package com.supermap.rnsupermap;

import com.facebook.react.bridge.Arguments;
import com.facebook.react.bridge.Promise;
import com.facebook.react.bridge.ReactApplicationContext;
import com.facebook.react.bridge.ReactContextBaseJavaModule;
import com.facebook.react.bridge.ReactMethod;
import com.facebook.react.bridge.WritableMap;
import com.supermap.analyst.BufferAnalystParameter;
import com.supermap.analyst.BufferEndType;
import com.supermap.analyst.BufferRadiusUnit;
import com.supermap.data.Enum;

import java.util.Calendar;
import java.util.HashMap;
import java.util.Map;

public class JSBufferAnalystParameter extends ReactContextBaseJavaModule {
    public static final String REACT_CLASS = "JSBufferAnalystParameter";
    protected static Map<String, BufferAnalystParameter> m_BufferAnalystParameterList = new HashMap<String, BufferAnalystParameter>();
    BufferAnalystParameter m_BufferAnalystParameter;

    public JSBufferAnalystParameter(ReactApplicationContext context) {
        super(context);
    }

    public static BufferAnalystParameter getObjFromList(String id) {
        return m_BufferAnalystParameterList.get(id);
    }

    @Override
    public String getName() {
        return REACT_CLASS;
    }

    public static String registerId(BufferAnalystParameter obj) {
        for (Map.Entry entry : m_BufferAnalystParameterList.entrySet()) {
            if (obj.equals(entry.getValue())) {
                return (String) entry.getKey();
            }
        }

        Calendar calendar = Calendar.getInstance();
        String id = Long.toString(calendar.getTimeInMillis());
        m_BufferAnalystParameterList.put(id, obj);
        return id;
    }

    @ReactMethod
    public void createObj(Promise promise){
        try{
            BufferAnalystParameter bufferAnalystParameter = new BufferAnalystParameter();
            String bufferAnalystParameterId = registerId(bufferAnalystParameter);

            WritableMap map = Arguments.createMap();
            map.putString("bufferAnalystParameterId",bufferAnalystParameterId);
            promise.resolve(map);
        }catch (Exception e){
            promise.reject(e);
        }
    }

    @ReactMethod
    public void setEndType(String bufferAnalystParameterId,int bufferEndType, Promise promise){
        try{
            BufferAnalystParameter bufferAnalystParameter = getObjFromList(bufferAnalystParameterId);
            bufferAnalystParameter.setEndType((BufferEndType) Enum.parse(BufferEndType.class,bufferEndType));

            promise.resolve(true);
        }catch (Exception e){
            promise.reject(e);
        }
    }

    @ReactMethod
    public void getEndType(String bufferAnalystParameterId, Promise promise){
        try{
            BufferAnalystParameter bufferAnalystParameter = getObjFromList(bufferAnalystParameterId);
            BufferEndType endType = bufferAnalystParameter.getEndType();
            int typeCode = Enum.getValueByName(BufferEndType.class,endType.name());
            WritableMap map = Arguments.createMap();
            map.putInt("EndType",typeCode);
            promise.resolve(map);
        }catch (Exception e){
            promise.reject(e);
        }
    }

    @ReactMethod
    public void setLeftDistance(String bufferAnalystParameterId,double distance,Promise promise){
        try{
            BufferAnalystParameter bufferAnalystParameter = getObjFromList(bufferAnalystParameterId);
            bufferAnalystParameter.setLeftDistance(distance);

            promise.resolve(true);
        }catch (Exception e){
            promise.reject(e);
        }
    }

    @ReactMethod
    public void getLeftDistance(String bufferAnalystParameterId,Promise promise){
        try{
            BufferAnalystParameter bufferAnalystParameter = getObjFromList(bufferAnalystParameterId);
            Object leftDistance =  bufferAnalystParameter.getLeftDistance();
            if (leftDistance instanceof String){
                WritableMap map = Arguments.createMap();
                map.putString("leftDistance",(String)leftDistance);
                promise.resolve(map);
            }else {
                WritableMap map = Arguments.createMap();
                map.putDouble("leftDistance",(double)leftDistance);
                promise.resolve(map);
            }
        }catch (Exception e){
            promise.reject(e);
        }
    }

    @ReactMethod
    public void setRightDistance(String bufferAnalystParameterId,double distance,Promise promise){
        try{
            BufferAnalystParameter bufferAnalystParameter = getObjFromList(bufferAnalystParameterId);
            bufferAnalystParameter.setRightDistance(distance);

            promise.resolve(true);
        }catch (Exception e){
            promise.reject(e);
        }
    }

    @ReactMethod
    public void getRightDistance(String bufferAnalystParameterId,Promise promise){
        try{
            BufferAnalystParameter bufferAnalystParameter = getObjFromList(bufferAnalystParameterId);
            Object rightDistance =  bufferAnalystParameter.getRightDistance();
            if (rightDistance instanceof String){
                WritableMap map = Arguments.createMap();
                map.putString("rightDistance",(String)rightDistance);
                promise.resolve(map);
            }else {
                WritableMap map = Arguments.createMap();
                map.putDouble("rightDistance",(double)rightDistance);
                promise.resolve(map);
            }
        }catch (Exception e){
            promise.reject(e);
        }
    }

    @ReactMethod
    public void setRadiusUnit(String bufferAnalystParameterId,int unit,Promise promise){
        try{
            BufferAnalystParameter bufferAnalystParameter = getObjFromList(bufferAnalystParameterId);
            bufferAnalystParameter.setRadiusUnit((BufferRadiusUnit) Enum.parse(BufferRadiusUnit.class,unit));

            promise.resolve(true);
        }catch (Exception e){
            promise.reject(e);
        }
    }

    @ReactMethod
    public void getRadiusUnit(String bufferAnalystParameterId, Promise promise){
        try{
            BufferAnalystParameter bufferAnalystParameter = getObjFromList(bufferAnalystParameterId);
            BufferRadiusUnit unit = bufferAnalystParameter.getRadiusUnit();
            int typeCode = Enum.getValueByName(BufferRadiusUnit.class,unit.name());
            WritableMap map = Arguments.createMap();
            map.putInt("radiusUnit",typeCode);
            promise.resolve(map);
        }catch (Exception e){
            promise.reject(e);
        }
    }

    @ReactMethod
    public void setSemicircleLineSegment(String bufferAnalystParameterId,int value,Promise promise){
        try{
            BufferAnalystParameter bufferAnalystParameter = getObjFromList(bufferAnalystParameterId);
            bufferAnalystParameter.setSemicircleLineSegment(value);

            promise.resolve(true);
        }catch (Exception e){
            promise.reject(e);
        }
    }

    @ReactMethod
    public void getSemicircleLineSegment(String bufferAnalystParameterId,Promise promise){
        try{
            BufferAnalystParameter bufferAnalystParameter = getObjFromList(bufferAnalystParameterId);
            int value = bufferAnalystParameter.getSemicircleLineSegment();

            WritableMap map = Arguments.createMap();
            map.putInt("segment",value);
            promise.resolve(map);
        }catch (Exception e){
            promise.reject(e);
        }
    }
}

